### PR TITLE
Feature/keeper incentives

### PR DESCRIPTION
**Pull request by RedVeil**
Created Contract for KeeperIncentives which can be inherited by multiple contracts. 
Each Function that should get keeper incentives must be individually initialized in KeeperIncentive.
When a Keeper calls the function in time they get rewards and the new time window for keeper rewards changes by a certain interval.

This is integrated now only in RewardsManager for swapTokenForRewards() and distributeRewards().

### CODE REVIEW

**Comment by anthonymartin** [comment]
@RedVeil should this be targeted to another branch? It looks like it has a lot of commits from another WIP branch

**Comment by RedVeil** [comment]
it looks like i used refactor-grant-elections for this. refactor-grant-elections also targets develop so i think this is ok.

**Comment by anthonymartin** [comment]
@RedVeil - when I target the PR to refactor-grant-elections there are conflicts displayed. also I see diffs with GrantElections that have nothing to do with this PR and that I would expect to be in `refactor-grant-elections`. e.g.  there are code changes related to randomNumber and vault share types in GrantElections. Could you help me out here?

File: packages/contracts/contracts/KeeperIncentive.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    bool enabled;
+    bool openToEveryone; //can everyone call the function to get the reward or only whitelisted?
+  }
+
+  /* ========== STATE VARIABLES ========== */
+
+  IERC20 public immutable POP;
+  Incentive[] public incentives;
+  uint256 public incentiveBudget;
+  mapping(address => bool) public whitelisted;
```
[the remark this answers is not in the document]
Comment (RedVeil) [reply packages/contracts/contracts/KeeperIncentive.sol]: its allowed now

**Comment by RedVeil** [comment]
Simplified the contract again.
Start, End and interval got removed so an incentive just contains the rewardAmount in POP, if its enabled and if everyone can call it.

File: packages/contracts/test/KeeperIncentive.test.ts [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+          .defaultIncentivisedFunction();
+
+        expect(result)
+          .to.emit(keeperIncentiveHelper, "FunctionCalled")
+          .withArgs(nonOwner.address);
+        const newbalance = await mockPop.balanceOf(nonOwner.address);
+        expect(newbalance).to.equal(oldBalance.add(incentive));
+      });
+    });
+    context("should not do anything ", function () {
```
Comment (anthonymartin) [inline packages/contracts/test/KeeperIncentive.test.ts:234]: the `context` and `it` statements here should be reversed

**Review by anthonymartin — APPROVED** [review]

**Pull request merged**